keps/sig-testing/20190605-presubmitconfig-in-tested-repo.md

```diff
@@ -0,0 +1,259 @@
+validation funcs for presubmits so that in the end a func can be
+exported that will be used both in case of config reload and
+when additional presubmits are loaded from a `prow.yaml`.
+
+*Open question:* What would be the best signature for such a function?
+It must be able to optionally accept additional presubmits for one repo
+
+#### New `prow/inrepoconfig/api` package
+
+A new package `prow/inrepoconfig/api` will be added which contains all
+the functionality needed by everything that wants to make use of the
+presubmits in `prow.yaml`. It will
+
+* Export a `type InRepoConfig` that will be used to unmarshal the contents of `prow.yaml` into
+* Export a `func New(log *logrus.Entry, c *config.Config, gc *git.Client, org, repo, baseRef string, headRefs []string) (*InRepoConfig, error)`
+
+The `func New` uses the passed in `*git.Client` to fetch the `baseRef`
+and all `headRefs` of the repo and merges them together using the
+appropriate strategy. The appropriate strategy can be found out by leveraging `*Config.Tide.MergeMethod(org, repo)`.
+Using the `*git.Client` has the advantage that it can do caching, which
+is very important for bigger repositories.
+After the merge was done, it is checked if a `prow.yaml` exists.
+Its absence will not be considered an error and an empty `InrepoConfig`
+will get returned.
+If `prow.yaml` exists, it will be unmarshalled using `yaml.UnmarshalStrict`
+to avoid human errors like typos in map keys. If that was successful,
+all presubmits found will be defaulted and validated using the exported
+logic from the `prow/config` package.
+
+*Alternative:* The whole logic for fetching and parsing the `prow.yaml`
+could be put into the `prow/config` package
+
+#### New `prow/inrepoconfig` package
+
+A new package `prow/inrepoconfig` will be added that exports a
+`func HandlePullRequest(log *logrus.Entry, c *config.Config, ghc githubClient, gc *git.Client, pr github.PullRequest)`. It:
+* Is intended to be used by the `trigger` plugin only
+* Initially creates a github status context to prevent merges while
+  fetching and parsing of `prow.yaml` is still in progress
+* Fetches and parses the `prow.yaml` using the `prow/inrepoconfig/api`
+  package
+* Posts a comment to GitHub if an error occurs
+* Updates the status context to `failure` or `success` after the
+  `prow.yaml` parsing is done
+* Has functionality to remove status contexts that existed in an older
+  `baseRef` and got removed in a newer one. This could be done by:
+    * Recognizing status contexts created by Prow  using a combination
+      of a static list (e.G. `tide`) and for the remaining contexts by
+      comparing their `targetURL` with what is configured in Prow as
+      `jobURLPrefix`. If the `targetURL` starts with `jobURlPrefix`
+      and the context does not belong to one of the known presubmits,
+      the status context will be retired.
+    * Leveraging the `status-reconciler`. The `status-reconciler`
```

Comment:
When? On merge?

Reply:
As mentioned further down the status reconciler would have to get notified on Push Events to any branch that is the base of any pull request, then reconcile all pull requests whose base changed.

IMHO the first approach is simpler thought.